bmwqemu.pm

```diff
@@ -213,13 +213,13 @@ sub log_call (@args) {
     else {
         # key/value pairs
         my @result;
-        my $should_hide = grep { ($_ // '') eq 'secret' } @args;
+        my %args = @_;
         while (my ($key, $value) = splice(@args, 0, 2)) {
             if ($key =~ tr/0-9a-zA-Z_//c) {
                 # only quote if needed
                 $key = pp($key);
             }
-            $value = "[masked]" if ($key eq 'text' && $should_hide);
```

Comment:
I would suggest not doing this here. This is something which should be done inside the caller.

But if we wanna keep it. it looks better then before :+1: .

Reply:
hm, I didn't checked all `log_call()` calles. There might be some who use `text` as the key for secret value!
As I say, this is just bad code and should not be done in this function!

Reply:
Why would you want to handle that in the caller and not here?

Reply:
It was originally changed here: https://github.com/os-autoinst/os-autoinst/pull/2002/files#diff-3420c7e01f73caee7d97053af39990e3297daf42d06e0fc8848f78af1a5e009eL1464
and it was accepted, I'm just correcting it

Reply:
> Why would you want to handle that in the caller and not here?

Imagine code like this:
```perl
sub setup_secret {
    my (%args) = @_;
   bmwqemu::log_call(@_, secret => 1);
    
    my $my_super_secret_password_is = $args{psk};
}
```

I understand the `log_call()` function in that way, that it log parameters which are given. And there is just **no** format of them. 

A workaround for me could be something like:
```perl
sub setup_secret {
    my (%args) = @_;
   bmwqemu::log_call(@_, -mask => ['psk']);
    
    my $my_super_secret_password_is = $args{psk};
}
```
And hope that no one will use `-mask` as parameters for the calling function :).

Reply:
A other idea is to mask by value and not by `key` like:
```perl
sub setup_secret {
    my (%args) = @_;
   bmwqemu::log_call(@_, -mask-value => [$args{psk}]);
    
    my $my_super_secret_password_is = $args{psk};
}
```
and the code snippet of `log_call()`
```perl
sub log_call {
    my $mask_value = [];
    my @args;
    while (my $v = shift @_) {
       if($v eq '-mask-value' && @_ ){
            $mask_value = shift @_;
            $mask_value = [$mask_value] unless ref($mask_value) eq 'ARRAY';
       } else {
            push @args, $v;
       }
    }

    foreach my $v (@args){
        foreach my $mask (@$mask) {
              $v =~ s/\Q$mask\E/[masked]/g;
        }
    }

    .... # Print out the arguements
}

Reply:
@okurz @jlausuch wdyt?

Reply:
What is the advantage of your proposal vs giving `secret => 1`? In both cases you need to tell `log_call` that the message is a secret in some way, either `secret =>` or `-maskvalue`...

Reply:
The advantage is, that you tell `log_call` **what** is the secret and not just that there is one.

Reply:
I am not exactly sure I understand your implementation proposal but it sounds like a better approach to explicitly pass a secret string around in an explicitly called variable rather than passing a string or text around and relying on an additional boolean variable to be correctly evaluated

Reply:
> I am not exactly sure I understand your implementation proposal but it sounds like a better approach to explicitly pass a secret string around in an explicitly called variable rather than passing a string or text around and relying on an additional boolean variable to be correctly evaluated

Is this a question for me?

Reply:
My comment was a response to @cfconrad. That implicitly includes a suggestion to you that you could try to implement it like that, if you can :)

Reply:
Agreed with @cfconrad that he will provide a PR with his implementation so we can compare better an decide what works better for us.

Reply:
here are my 5cent of this: https://github.com/os-autoinst/os-autoinst/pull/2062